Support SET NO_BROSETABLE ON and OFF

ZhenyeLee opened the pull request:
Supprot  SET NO_BROWSETABLE ON/Off. For SET NO_BROWSETABLE ON and OFF, errors are raised, and escape hatch are created for ignoring the setting as well.

Task: BABEL-2939 https://jira.rds.a2z.com/browse/BABEL-2939
Signed-off-by: Zhenye Li <zhenye@amazon.com>

### Description

Task: BABEL-2939 https://jira.rds.a2z.com/browse/BABEL-2939
CR: https://code.amazon.com/reviews/CR-71928621 (merged)
 
### Issues Resolved

Supprot  SET NO_BROWSETABLE ON/Off. For SET NO_BROWSETABLE ON and OFF, errors are raised, and escape hatch are created for ignoring the setting as well. 

### Unit Testing
```
     [java] ###########################################################################
     [java] TOTAL TESTS:	655
     [java] TESTS PASSED:	655
     [java] TESTS FAILED:	0
     [java] ###########################################################################
```

### Check List
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Support SET NO_BROSETABLE ON and OFF

Supprot  SET NO_BROWSETABLE ON/Off. For SET NO_BROWSETABLE ON and OFF, errors are raised, and escape hatch are created for ignoring the setting as well.

Task: BABEL-2939
Signed-off-by: Zhenye Li <zhenye@amazon.com>

Commit:
Support SET NO_BROSETABLE ON and OFF

Supprot  SET NO_BROWSETABLE ON/Off. For SET NO_BROWSETABLE ON and OFF, errors are raised, and escape hatch are created for ignoring the setting as well.

Task: BABEL-2939
Signed-off-by: Zhenye Li <zhenye@amazon.com>

Commit:
Support SET NO_BROSETABLE ON and OFF

Supprot  SET NO_BROWSETABLE ON/Off. For SET NO_BROWSETABLE ON and OFF, errors are raised, and escape hatch are created for ignoring the setting as well.

Task: BABEL-2939
Signed-off-by: Zhenye Li <zhenye@amazon.com>

faviansamatha committed:
Fixed upgrade script regarding the following msdb policy objects: fn_syspolicy_is_automation_enabled, syspolicy_configuration, and syspolicy_system_health_state (#351)

Previously, several msdb objects were checked in (msdb.dbo.fn_syspolicy_is_automation_enabled(), msdb.dbo.syspolicy_configuration, and msdb.dbo.syspolicy_system_health_state) with an error in the upgrade script which did not assign sysadmin as the owner of these objects. Furthermore, sys.shipped_objects_not_in_sys needed to be modified in order to include these objects as well. This commit fixes both of those issues.

Tasks: BABELFISH-439, BABELFISH-441, BABELFISH-442
Signed-off-by: Favian (Ian) Samatha ians@bitquilltech.com

rossd-bitquill committed:
Add error to sp_describe_undeclared_parameters stored procedure for unsupported use case (#315)

This PR adds a new error message "Unsupported use case..." (error code 0A000) to the stored
procedure sp_describe_undeclared_parameters to avoid a possible segfault that could cause Babelfish to become unresponsive.

This error will occur if sp_describe_undeclared_parameters is called with more than one list of VALUES for an INSERT statement, e.g. sp_describe_undeclared_parameters @tsql = N'INSERT INTO mytable VALUES (@p1, @p2, @p3), (@P4, @p5, @p6)'.

Additionally, update the tests for sp_describe_undeclared_parameters to test this new unsupported use case error.

Task: BABELFISH-367

Signed-off-by: rossd-bitquill dutkiert@amazon.com

faviansamatha committed:
sys.triggers Implementation (#350)

This commit adds the sys.triggers view to Babelfish, which unblocks SSMS Generate Script Wizard functionalities.

NOTE: This currently only supports DML triggers, as DDL triggers are not yet supported.

Task: BABELFISH-479
Signed-off-by: Favian (Ian) Samatha <ians@bitquilltech.com>

Deepesh125 committed:
[BABEL-3316] fix tds_collation_id and tds_collation_sort_id fields of sp_describe_first_result_set procedure (#310)

BCP in functionality with any collations other than english collation
was failing because of incorrect configuration of tds_collation_id and
tds_collation_sort_id fields of sp_describe_first_result_set procedure
which BCP utility internally uses to determine metadata.

This commit fixes tds_collation_id and tds_collation_sort_id fields by
adding appropriate property support in COLLATIONPROPERTY() which will be
used by sp_describe_first_result_set.

Co-authored-by: Dipesh Dhameliya <dddhamel@amazon.com>

Deepesh125 committed:
Support multi-byte client encodings other than UTF-16 for VARCHAR(n)

For the non-Unicode datatypes CHAR, VARCHAR, or TEXT, "n" is interpreted as the maximum number of bytes that may be stored in the encoding of the collation. For example, this means that if a Japanese collation is used, then n is the maximum number of bytes in the SJIS encoding. PostgreSQL always stores strings in the server encoding, and always interprets "n" as the maximum number of characters, but it is possible for a string of n characters to require more than n bytes in the client encoding.

With this change, byte length of input would be validated at insertion or update time. And appropriate warning or error would be raised if needed.

Also, moved encoding conversion functionality from babelfishpg_tds to babelfishpg_tsql because we need byte length checking logic in TSQL dialect also irrespective of whether client is TDS.

Task: #338 
BABEL-3090
Signed-off-by: Dipesh Dhameliya <dddhamel@amazon.com>
Authored-by: Kuntal Ghosh <kuntalgh@amazon.com>

sangilsong commented on contrib/babelfishpg_tsql/src/pltsql_instr.h:
```diff
@@ -636,5 +636,6 @@ typedef enum PgTsqlInstrMetricType {
 
-	INSTR_TSQL_COUNT
+	INSTR_TSQL_COUNT,
+	INSTR_UNSUPPORTED_TSQL_OPTION_NO_BROWSETABLE
```
new enum should be inserted before INSTR_TSQL_COUNT

Deepesh125 committed:
Move dependency of collation from babelfishpg_tsql to babelfishpg_common extension

- Move all collation related arrays and hash table to babelfishpg_common extension because collations are essential in order to work babelfish defined datatypes correctly e.g., sys."varchar" and sys."bpchar".

- In most of the cases, we don't need to maintain separate table for lcid to encoding mapping on TDS except a few use cases, for example, prepare-execute where we need to find encoding based on lcid. Also, We were doing extra hash lookup to find encoding based on lcid once we find lcid from collation oid. In fact, we can also get encoding from collation oid directly. So we can save extra lookup for encoding by combining lcid to encoding mapping data with existing coll_info table.

- Removed "PLtsql_collation_callbacks" from babelfishpg_tsql extension. We introduced this collation callback in new for "collation_callbacks" to avoid dependency of pltsql.h from babelfishpg_common extension. babelfishpg_tsql extension is now dependent on babelfishpg_common for collation related functionalities.

- Move encoding conversion functionalities to babelfishpg_common extension. We had to do this because datatypes input function for sys."varchar" and sys."bpchar" required these encoding conversion functionalities to calculate the byte length of the input.

Task: #360 
BABEL-3255, BABEL-3356
Authored-by: Dipesh Dhameliya <dddhamel@amazon.com>
Signed-off-by: Kuntal Ghosh <kuntalgh@amazon.com>
